### ISSUE TITLE
`ref_as_ptr` example does not emit warnings

### ISSUE THREAD

**Issue by y21** [rust-lang/rust-clippy#12771]
### Summary

The example in https://rust-lang.github.io/rust-clippy/master/index.html#/ref_as_ptr does not emit any warnings for the lint.

### Lint Name

ref_as_ptr

### Reproducer

I tried this code:

```rust
#![warn(clippy::ref_as_ptr)]
fn main() {
  let a_ref = &1i32;
  let a_ptr = a_ref as *const _;
}
```

I expected to see this happen:
Since this is the example for what not to do, the lint should emit a warning here suggesting to use `ptr::from_ref`.

Instead, this happened:
No warnings.

### Version

```text
clippy commit befb659145a734e51a3708ff84315ae07123c880
```

**Comment by J-ZhengLi** [comment]
related PR: https://github.com/rust-lang/rust-clippy/pull/12260

the doc should be updated

**Comment by briansmith** [comment]
I don't think this is a documentation issue. PR #12260 overcorrected and isn't using the right criteria to detect lifetime extension.
```
fn foo<T>(mut x: T) {
  {
    let r = &mut x;
    let _p = r as *mut T; // <- The lint should definitely trigger here; no lifetime extension.
  }
  {
    let r = &x;
    let _p = r as *const T; // <- The lint should definitely trigger here; no lifetime extension.
  }
  {
    let _p = &mut x as *mut T; // <- The lint should probably trigger here since `x` isn't temporary.
  }
  {
    let _p = &x as *const T; // <- The lint should probably trigger here since `x` isn't temporary.
  }
}
```
